### PR TITLE
Creation of InputTextArea component

### PR DESCRIPTION
**Pull request by Valerian-Perez-Wanadev**
Based on fneitzel prototype ([Added prototype of InputTextArea](https://github.com/BabylonJS/Babylon.js/pull/8275)), I improved the component to support the margin, the deletion (backspace/del.), the highlighting, deadkeys and copy/paste/cut shortcuts.

A lot of things remained to do :
- fix the CR on last line
- find why the text clip on single line during autostreching
- resolve todos
- clean/fix/update the code

My approach was highly naïve, so maybe a cleaner architecture will be necessery.

### CODE REVIEW

**Comment by fneitzel** [comment]
Cool that you are finishing it, let me know if you need help.

File: gui/src/2D/controls/inputTextArea.ts
Comment (RaananW) [inline gui/src/2D/controls/inputTextArea.ts:18]: nit - why id you end with an underscore?

File: gui/src/2D/controls/inputTextArea.ts
Comment (RaananW) [inline gui/src/2D/controls/inputTextArea.ts:32]: Is there a reason for these values? why use "-1"? and why doesn't ellipsis have any value?

File: gui/src/2D/controls/inputTextArea.ts
Comment (RaananW) [inline gui/src/2D/controls/inputTextArea.ts:40]: would be great to cleanup any unneeded / commented line of code

File: gui/src/2D/controls/inputTextArea.ts
Comment (RaananW) [inline gui/src/2D/controls/inputTextArea.ts:63]: nit - naming. Measure might be confusing (as it is probably thought to be the measure of the control). maybe "textMeasurement"?

File: gui/src/2D/controls/inputTextArea.ts
Comment (RaananW) [inline gui/src/2D/controls/inputTextArea.ts:138]: why is this not a part of the switch?

File: gui/src/2D/controls/inputTextArea.ts
Comment (RaananW) [inline gui/src/2D/controls/inputTextArea.ts:1048]: I understand why this is needed, but this could be solved within the render loop using timers - 
https://doc.babylonjs.com/divingDeeper/events/observables#observable-based-countdown-function
setTimeout could cause unexpected results, as it will surely run outside of the render loop

**Review by RaananW — COMMENTED** [review]
Wow, thank you so much for the. Such a wonderful addition to the GUI!

I have a few comments. The code should be cleaned and formatted, but those are minor issues.

My bigger question is this - I noticed you implemented yourself backspace, delete and so on. This is, eventually, a 3d-controlled <textarea> element. Why not use a textarea as the backend of this component and allow the browser to do the heavy-lifting of text editing?

**Comment by RaananW** [comment]
I also wait for @DarraghBurkeMS's review to test the actual functionality.

**Comment by Popov72** [comment]
> Why not use a textarea as the backend of this component and allow the browser to do the heavy-lifting of text editing?

What about Native if using a textarea HTML element under the hood?

**Comment by RaananW** [comment]
> What about Native if using a textarea HTML element under the hood?

That's a good point. @bghgary , is there some corresponding element to be used, or does the functionality needs to be (re) implemented?

**Comment by darraghjburke** [comment]
I'm not Gary, but: I don't think we can depend on any external text area implementation in Babylon Native. One of the critical design choices is that it's "bring your own UI", so we can't rely on any particular one of the many cross-platform UI solutions out there. The goal AFAIK is to implement enough of the canvas API that we can support Babylon GUI, but Babylon Native has no plans to support any DOM features (at that point we'd be making a browser.)

From my POV, having a Babylon GUI text area brings a lot of value, but it's also worth keeping in mind that it will probably be significantly harder to maintain than our existing controls. That being said, I love that you've taken a stab at it, and I'll definitely give it a try later this week :)

**Comment by deltakosh** [comment]
Totally aligned with @darraghjburke : no external dependencies beside just canvas API

**Comment by bghgary** [comment]
> What about Native if using a textarea HTML element under the hood?

Maybe @CedricGuillemet can answer this part.

In general, I'm also curious if there are any localization concerns with this change.

**Comment by Valerian-Perez-Wanadev** [comment]
Thanks a lot for all your feedbacks, I'll fix that later in the week!

**Comment by CedricGuillemet** [comment]
> > What about Native if using a textarea HTML element under the hood?
> 
> Maybe @CedricGuillemet can answer this part.
> 

Hard to tell without testing. At the current state, (basic) text and shape rendering will work. Idk if any unimplemented method is needed here.

**Comment by deltakosh** [comment]
@Valerian-Perez-Wanadev : @fneitzel is interested to contribute to your PR. Do you mind opening your branch for his PR?

File: gui/src/2D/controls/inputTextArea.ts
Comment (Valerian-Perez-Wanadev) [reply gui/src/2D/controls/inputTextArea.ts]: It's an old implementation of first prototype. I removed it and reimported from the correct file.

File: gui/src/2D/controls/inputTextArea.ts
Comment (Valerian-Perez-Wanadev) [reply gui/src/2D/controls/inputTextArea.ts]: It was certainly based on the enum in `textBlock.ts:14` which has the same behaviour. But I don't know why "-1" was used.

File: gui/src/2D/controls/inputTextArea.ts
Comment (Valerian-Perez-Wanadev) [reply gui/src/2D/controls/inputTextArea.ts]: Removed!

File: gui/src/2D/controls/inputTextArea.ts
Comment (Valerian-Perez-Wanadev) [reply gui/src/2D/controls/inputTextArea.ts]: Correct. It's changed to `textMetrics` to be complient with the type.

File: gui/src/2D/controls/inputTextArea.ts
Comment (Valerian-Perez-Wanadev) [reply gui/src/2D/controls/inputTextArea.ts:1019]: Oh, great!

Fneitzel uses the same method as in `inputText.ts`, but I can improve it today.

File: gui/src/2D/controls/inputTextArea.ts
Comment (Valerian-Perez-Wanadev) [reply gui/src/2D/controls/inputTextArea.ts]: Oh, yes.

File: gui/src/2D/controls/inputTextArea.ts
Comment (Valerian-Perez-Wanadev) [reply gui/src/2D/controls/inputTextArea.ts:1019]: Currently, I'm facing some issues over AdvancedTimer class. 
- Is it correct to instantiate it in `_link` function to have access to the scene ?
- I can't stop/start in the same loop to reset the timer. 

I want to avoid SetAndStart to save some perf but it becomes to be the only solution...
Have you some ideas to bypass this ?

**Comment by Valerian-Perez-Wanadev** [comment]
I've made some good improvments during the sart of the year but I'm in a dilemma...

The Component inherits from InputText, but it acts like a TextBlock. So for the textWrapping behaviour, I must follow the implementation of only Clip/WordWrap or Clip/WordWrap/Ellipsis/WordWrapEllipsis?

Furthermore, the ellipsis mode it's not very useful in InputTextArea... I choose to discard it for the moment. But I want to known your thoughts about this before merging my code.

**Comment by deltakosh** [comment]
For things in common we can think about having a common ancestor?

**Comment by Valerian-Perez-Wanadev** [comment]
> For things in common we can think about having a common ancestor?

Of course I will do thatif needed but actually my choice is on which behaviour to adopt. 

After seeing implementation of TextArea of Firefox and the InputText of BJS, I will keep their philosophy and let the user decides whether their add a line or let the component breaks the text alone. It's more natural to do it that way.

**Comment by Valerian-Perez-Wanadev** [comment]
I have a question about the Enter key.

How can I override the blur by hitting Enter ? 

I want to add a new line with Enter but I can't figure how to do without blurring the input:/

**Comment by deltakosh** [comment]
the onkeydown can be cancelled if you detect Enter (preventDefault)

**Comment by dad72** [comment]
This a very useful feature. Thank you. I can't wait to use it to replace what I did in HTML

**Comment by Valerian-Perez-Wanadev** [comment]
It's almost done. Some bugs are complicated to fix but I'm always working on it :)

**Comment by jonathanandrewsuk** [comment]
@darraghjburke @RaananW just bumping this to register my appreciation for this PR!

File: packages/dev/gui/src/2D/controls/inputTextArea.ts
Comment (RaananW) [inline packages/dev/gui/src/2D/controls/inputTextArea.ts:180]: keyCode was already deprecated and will slowly be removed from browsers. And we don't support IE11 anymore (bless whoever's in charge of that :-)) - https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode

Would it be possible to work with `key` or `code` instead? Or was there any specific reason for choosing to work with keyCode?

I understand we use this internally quite often. And we need to change that (sooner rather than later). I am just trying to avoid commiting new code with deprecated APIs.

File: packages/dev/gui/src/2D/controls/inputTextArea.ts
Comment (RaananW) [inline packages/dev/gui/src/2D/controls/inputTextArea.ts:1210]:
```suggestion
    protected _processDblClick(_evt: PointerInfo) {
```

(or remove the variable altogether :-) )

**Review by RaananW — COMMENTED** [review]
This looks great!!

I just have a question about keyCode vs. key (and just a small suggestion that can be ignored).
Is it possible to see a working playground with this in the snapshot playground?

**Comment by matthewpflueger** [comment]
Would love to see this merged! This is a perfect fit for our project...

File: packages/dev/gui/src/2D/controls/inputTextArea.ts
Comment (Valerian-Perez-Wanadev) [reply packages/dev/gui/src/2D/controls/inputTextArea.ts]: I prefer to keep the parameter, it will be more maintainable after the release (just in case).

File: packages/dev/gui/src/2D/controls/inputTextArea.ts
Comment (Valerian-Perez-Wanadev) [reply packages/dev/gui/src/2D/controls/inputTextArea.ts:180]: Hmm, I didn't notice that... I don't have any reason to keep this deprecated API. I will rewrite the processkey ASAP 👍

Comment (Valerian-Perez-Wanadev) [reply packages/dev/gui/src/2D/controls/inputTextArea.ts:180]: @RaananW, unfortnaly I can't reproduce a correct behavior without the deprecation. The charCode filters the character and discard all modifier key like `Shift`, `Alt`, etc.

Actually, I can achieve almost the same things but I'm stuck to allow two duet characters: `^^` and ` `` `

File: gui/src/2D/controls/inputTextArea.ts
Comment (Valerian-Perez-Wanadev) [reply gui/src/2D/controls/inputTextArea.ts]: I can't restart the timer during the same frame which it's stop

File: packages/dev/gui/src/2D/controls/inputTextArea.ts
Comment (Valerian-Perez-Wanadev) [reply packages/dev/gui/src/2D/controls/inputTextArea.ts:180]: @RaananW, Can I push this evol?

**Review by RaananW — APPROVED** [review]

File: packages/dev/gui/src/2D/controls/inputTextArea.ts
Comment (RaananW) [reply packages/dev/gui/src/2D/controls/inputTextArea.ts]: Can this be resolved, or do we have some more to discuss?

Comment (Valerian-Perez-Wanadev) [reply packages/dev/gui/src/2D/controls/inputTextArea.ts]: We can resolve of course! :)

**Comment by Valerian-Perez-Wanadev** [comment]
@RaananW  Just one thing before merging. It's possible to assign @fneitzel as co-author on the changelog please? Originally it was his code and we work sometimes together on this final version.

**Comment by RaananW** [comment]
> @RaananW Just one more thing before merging. It's possible to assign @fneitzel as co-author on the changelog please? Originally it was his code and we work sometimes together on this final version.

I will have to edit the changelog afterwards, as it is generated automatically, but i'd be happy to do that. Please remind me after the next version is merged :-)


Also a request regarding this PR - we will need to update the documentation with this new feature. Adding one or two working examples will also be wonderful. Will you be able to do that?

**Comment by Valerian-Perez-Wanadev** [comment]
> > @RaananW Just one more thing before merging. It's possible to assign @fneitzel as co-author on the changelog please? Originally it was his code and we work sometimes together on this final version.
> 
> I will have to edit the changelog afterwards, as it is generated automatically, but i'd be happy to do that. Please remind me after the next version is merged :-)
> 
> Also a request regarding this PR - we will need to update the documentation with this new feature. Adding one or two working examples will also be wonderful. Will you be able to do that?

I'll do that 👍 

And my local PG has some beautiful examples!

**Comment by jonathanandrewsuk** [comment]
@Valerian-Perez-Wanadev I'd be happy to help out with the docs (I've been using the component for a few weeks already)

**Pull request merged**
